libyul/optimiser/RedundantStoreEliminator.h

```diff
@@ -0,0 +1,117 @@
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * Optimiser component that removes stores to memory and storage slots that are not used
+ * or overwritten later on.
+ */
+
+#pragma once
+
+#include <libyul/ASTForward.h>
+#include <libyul/optimiser/ASTWalker.h>
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libyul/optimiser/Semantics.h>
+#include <libyul/optimiser/RedundantStoreBase.h>
+
+#include <map>
+#include <vector>
+
+namespace solidity::yul
+{
+struct Dialect;
+struct AssignedValue;
+
+/**
+ * Optimizer component that removes mstore, sstore and similar statements if they
+ * are overwritten in all code paths or never read from.
+ *
+ * The m_store member of RedundantStoreBase is only used with the empty yul string
+ * as key in the first dimension.
+ *
+ * Best run in SSA form.
+ *
+ * Prerequisite: Disambiguator, ForLoopInitRewriter.
+ */
+class RedundantStoreEliminator: public RedundantStoreBase
+{
+public:
+	static constexpr char const* name{"RedundantStoreEliminator"};
+	static void run(OptimiserStepContext& _context, Block& _ast);
+
+	explicit RedundantStoreEliminator(
+		Dialect const& _dialect,
+		std::map<YulString, SideEffects> const& _functionSideEffects,
+		std::map<YulString, ControlFlowSideEffects> _controlFlowSideEffects,
+		std::map<YulString, AssignedValue> const& _ssaValues,
+		bool _ignoreMemory
+	):
+		RedundantStoreBase(_dialect),
+		m_ignoreMemory(_ignoreMemory),
+		m_functionSideEffects(_functionSideEffects),
+		m_controlFlowSideEffects(_controlFlowSideEffects),
+		m_ssaValues(_ssaValues)
+	{}
+
+	using RedundantStoreBase::operator();
+	void operator()(FunctionCall const& _functionCall) override;
+	void operator()(FunctionDefinition const&) override;
+	void operator()(Leave const&) override;
+
+	using RedundantStoreBase::visit;
+	void visit(Statement const& _statement) override;
+
+	enum class Location { Storage, Memory };
+	enum class Effect { Read, Write };
+	struct Operation
+	{
+		Location location;
+		Effect effect;
+		/// Start of affected area. Unknown if not provided.
+		std::optional<YulString> start;
+		/// Length of affected area, unknown if not provided.
+		/// Unused for storage.
+		std::optional<YulString> length;
+	};
+
+private:
+	void shortcutNestedLoop(TrackedStores const&) override
+	{
+		// We might only need to do this for newly introduced stores in the loop.
+		changeUndecidedTo(State::Used);
+	}
+	void finalizeFunctionDefinition(FunctionDefinition const&) override { scheduleUnusedForDeletion(); }
```

Comment:
What happens to the statements that are undecided when we reach the end of a function definition? This looks like a place where they should be handled but this code does not do anything about them.

Is that because we consider them "used" so they would not be deleted anyway? If that's the case then I'd at least add a comment here saying so.

Reply:
It is done at the very end of the optimizer step, but it's probably good to do it here, too.

Reply:
Ah no, this is actually a bug!

Reply:
Yeah, you are right, probably it does not matter because everything that is not marked "unused" does not matter, but I'll add it anyway.